File: mtgSearch/src/main/java/com/dbottillo/mtgsearchfree/view/fragments/BasicFragment.java
package com.dbottillo.mtgsearchfree.view.fragments;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;

import com.dbottillo.mtgsearchfree.MTGApp;
import com.dbottillo.mtgsearchfree.util.LOG;
import com.dbottillo.mtgsearchfree.util.LeakCanaryUtil;
import com.dbottillo.mtgsearchfree.util.TrackingManager;
import com.dbottillo.mtgsearchfree.view.activities.BasicActivity;

public abstract class BasicFragment extends DialogFragment {

    protected BasicActivity dbActivity;
    protected boolean isPortrait = false;
    protected MTGApp app;

    public static final String PREF_SHOW_IMAGE = "show_image";
    public static final String PREF_SCREEN_ON = "screen_on";
    public static final String PREF_TWO_HG_ENABLED = "two_hg";
    public static final String PREF_SORT_WUBRG = "sort_wubrg";


    public void onAttach(Context context) {
        super.onAttach(context);
        LOG.d();

        this.dbActivity = (BasicActivity) context;
        app = (MTGApp) dbActivity.getApplication();
        isPortrait = getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT;
    }

    protected MTGApp getMTGApp() {
        return dbActivity.getMTGApp();
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        LOG.d();

        setHasOptionsMenu(true);
    }


    public void onDestroy() {
        super.onDestroy();
        LeakCanaryUtil.watchFragment(this);
    }

    void setActionBarTitle(String title) {
        if (dbActivity.getSupportActionBar() != null) {
            dbActivity.getSupportActionBar().setTitle(title);
        }
    }

    public void onResume() {
        super.onResume();
        LOG.d();
        TrackingManager.trackPage(getPageTrack());
    }

    public abstract String getPageTrack();


    public boolean onBackPressed() {
        return false;
    }

}

File: mtgSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/view/fragments/BasicFragment.kt
package com.dbottillo.mtgsearchfree.view.fragments

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.res.Configuration
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v7.app.AppCompatActivity
import com.dbottillo.mtgsearchfree.base.MTGApp
import com.dbottillo.mtgsearchfree.communication.events.BaseEvent
import com.dbottillo.mtgsearchfree.component.AppComponent
import com.dbottillo.mtgsearchfree.helper.LOG
import com.dbottillo.mtgsearchfree.helper.TrackingHelper
import de.greenrobot.event.EventBus

abstract class BasicFragment : DialogFragment() {

    protected var dbActivity: AppCompatActivity? = null
    protected var isPortrait: Boolean = false
    protected var bus = EventBus.getDefault()

    companion object {
        val PREF_SHOW_IMAGE = "show_image"
        val PREF_SCREEN_ON = "screen_on"
        val PREF_TWO_HG_ENABLED = "two_hg"
        val PREF_SORT_WUBRG = "sort_wubrg"
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        this.dbActivity = context as AppCompatActivity?
        val res = context!!.resources
        isPortrait = res.configuration.orientation == Configuration.ORIENTATION_PORTRAIT
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        //setupComponent(MTGApp.Companion.graph);

        setHasOptionsMenu(true)
    }

    override fun onPause() {
        super.onPause()
        bus.unregister(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        val refWatcher = MTGApp.refWatcher
        refWatcher?.watch(this)
    }

    val sharedPreferences: SharedPreferences
        get() = dbActivity!!.getSharedPreferences(MTGApp.Companion.PREFS_NAME, 0)

    protected fun setActionBarTitle(title: String) {
        if (dbActivity!!.supportActionBar != null) {
            dbActivity!!.supportActionBar!!.title = title
        }
    }

    protected fun openPlayStore() {
        TrackingHelper.getInstance(dbActivity!!.applicationContext).trackEvent(TrackingHelper.UA_CATEGORY_POPUP, TrackingHelper.UA_ACTION_OPEN, "play_store")
        val appPackageName = "com.dbottillo.mtgsearch"
        try {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + appPackageName)))
        } catch (anfe: ActivityNotFoundException) {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("http://play.google.com/store/apps/details?id=" + appPackageName)))
        }

    }

    override fun onResume() {
        super.onResume()
        if (getPageTrack() != null) {
            TrackingHelper.getInstance(dbActivity!!.applicationContext).trackPage(getPageTrack())
        }
        bus.registerSticky(this)
    }

    abstract fun getPageTrack(): String?

    val app: MTGApp
        get() = dbActivity!!.application as MTGApp

    open fun onBackPressed(): Boolean {
        return false
    }

    fun onEvent(event: BaseEvent<Any>) {

    }

    fun getIsPortrait(): Boolean {
        return isPortrait;
    }

    abstract fun setupComponent(appComponent: AppComponent)

}
